File: src/cloudtwin/providers/gcp/pubsub/handlers.py
"""
GCP Pub/Sub HTTP handlers – Pub/Sub REST API v1.

Routes match what PublisherRestTransport and SubscriberRestTransport send:

  Topics:
    PUT    /v1/projects/{project}/topics/{topic}     → create topic
    GET    /v1/projects/{project}/topics/{topic}     → get topic
    GET    /v1/projects/{project}/topics             → list topics
    DELETE /v1/projects/{project}/topics/{topic}     → delete topic
    POST   /v1/projects/{project}/topics/{topic}:publish  → publish

  Subscriptions:
    PUT    /v1/projects/{project}/subscriptions/{sub}     → create subscription
    GET    /v1/projects/{project}/subscriptions/{sub}     → get subscription
    GET    /v1/projects/{project}/subscriptions           → list subscriptions
    DELETE /v1/projects/{project}/subscriptions/{sub}     → delete subscription
    POST   /v1/projects/{project}/subscriptions/{sub}:pull          → pull
    POST   /v1/projects/{project}/subscriptions/{sub}:acknowledge   → acknowledge
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, Request, Response
from fastapi.responses import JSONResponse

from cloudtwin.core.errors import NotFoundError
from cloudtwin.providers.gcp.pubsub.service import PubsubService

log = logging.getLogger("cloudtwin.gcp.pubsub")


def _topic_json(t) -> dict:
    return {"name": t.full_name}


def _sub_json(s) -> dict:
    return {
        "name": s.full_name,
        "topic": s.topic_full_name,
        "ackDeadlineSeconds": s.ack_deadline_seconds,
        "messageRetentionDuration": "604800s",
    }


def make_router(service: PubsubService) -> APIRouter:
    router = APIRouter()

    # ------------------------------------------------------------------
    # Topics
    # ------------------------------------------------------------------

    @router.put("/v1/projects/{project}/topics/{topic}")
    async def create_topic(project: str, topic: str) -> JSONResponse:
        full_name = f"projects/{project}/topics/{topic}"
        t = await service.create_topic(full_name)
        return JSONResponse(_topic_json(t), status_code=200)

    @router.get("/v1/projects/{project}/topics/{topic}")
    async def get_topic(project: str, topic: str) -> JSONResponse:
        full_name = f"projects/{project}/topics/{topic}"
        try:
            t = await service.get_topic(full_name)
        except NotFoundError:
            return JSONResponse(
                {"error": {"code": 404, "message": f"Topic {full_name!r} not found"}},
                status_code=404,
            )
        return JSONResponse(_topic_json(t))

    @router.get("/v1/projects/{project}/topics")
    async def list_topics(project: str) -> JSONResponse:
        topics = await service.list_topics(project=project)
        return JSONResponse({"topics": [_topic_json(t) for t in topics]})

    @router.delete("/v1/projects/{project}/topics/{topic}")
    async def delete_topic(project: str, topic: str) -> Response:
        full_name = f"projects/{project}/topics/{topic}"
        try:
            await service.delete_topic(full_name)
        except NotFoundError:
            return JSONResponse(
                {"error": {"code": 404, "message": f"Topic {full_name!r} not found"}},
                status_code=404,
            )
        return JSONResponse({})

    @router.post("/v1/projects/{project}/topics/{topic}:publish")
    async def publish(project: str, topic: str, request: Request) -> JSONResponse:
        full_name = f"projects/{project}/topics/{topic}"
        body = await request.json()
        messages = body.get("messages", [])
        try:
            message_ids = await service.publish(full_name, messages)
        except NotFoundError:
            return JSONResponse(
                {"error": {"code": 404, "message": f"Topic {full_name!r} not found"}},
                status_code=404,
            )
        return JSONResponse({"messageIds": message_ids})

    # ------------------------------------------------------------------
    # Subscriptions
    # ------------------------------------------------------------------

    @router.put("/v1/projects/{project}/subscriptions/{sub}")
    async def create_subscription(
        project: str, sub: str, request: Request
    ) -> JSONResponse:
        full_name = f"projects/{project}/subscriptions/{sub}"
        body = await request.json()
        topic_full_name = body.get("topic", "")
        ack_deadline = body.get("ackDeadlineSeconds", 10)
        try:
            s = await service.create_subscription(
                full_name, topic_full_name, ack_deadline_seconds=ack_deadline
            )
        except NotFoundError as exc:
            return JSONResponse(
                {"error": {"code": 404, "message": exc.message}}, status_code=404
            )
        return JSONResponse(_sub_json(s), status_code=200)

    @router.get("/v1/projects/{project}/subscriptions/{sub}")
    async def get_subscription(project: str, sub: str) -> JSONResponse:
        full_name = f"projects/{project}/subscriptions/{sub}"
        try:
            s = await service.get_subscription(full_name)
        except NotFoundError:
            return JSONResponse(
                {
                    "error": {
                        "code": 404,
                        "message": f"Subscription {full_name!r} not found",
                    }
                },
                status_code=404,
            )
        return JSONResponse(_sub_json(s))

    @router.get("/v1/projects/{project}/subscriptions")
    async def list_subscriptions(project: str) -> JSONResponse:
        subs = await service.list_subscriptions(project=project)
        return JSONResponse({"subscriptions": [_sub_json(s) for s in subs]})

    @router.delete("/v1/projects/{project}/subscriptions/{sub}")
    async def delete_subscription(project: str, sub: str) -> Response:
        full_name = f"projects/{project}/subscriptions/{sub}"
        try:
            await service.delete_subscription(full_name)
        except NotFoundError:
            return JSONResponse(
                {
                    "error": {
                        "code": 404,
                        "message": f"Subscription {full_name!r} not found",
                    }
                },
                status_code=404,
            )
        return JSONResponse({})

    @router.post("/v1/projects/{project}/subscriptions/{sub}:pull")
    async def pull(project: str, sub: str, request: Request) -> JSONResponse:
        full_name = f"projects/{project}/subscriptions/{sub}"
        body = await request.json()
        max_messages = body.get("maxMessages", 10)
        try:
            received = await service.pull(full_name, max_messages=max_messages)
        except NotFoundError:
            return JSONResponse(
                {
                    "error": {
                        "code": 404,
                        "message": f"Subscription {full_name!r} not found",
                    }
                },
                status_code=404,
            )
        return JSONResponse({"receivedMessages": received})

    @router.post("/v1/projects/{project}/subscriptions/{sub}:acknowledge")
    async def acknowledge(project: str, sub: str, request: Request) -> Response:
        full_name = f"projects/{project}/subscriptions/{sub}"
        body = await request.json()
        ack_ids = body.get("ackIds", [])
        await service.acknowledge(full_name, ack_ids)
        return JSONResponse({})

    return router

File: rust/cloudtwin-lite/src/gcp/pubsub/handlers.rs
//! GCP Pub/Sub HTTP handlers.
//!
//! These map Google-style REST endpoints onto the provider service methods and
//! keep the request/response JSON glue out of `service.rs`.

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, put},
    Json, Router,
};

use super::service::PubsubService;
use crate::AppState;

fn svc(state: &Arc<AppState>) -> PubsubService {
    PubsubService::new(state.db.clone(), state.cfg.gcp_project.clone())
}

/// Router is mounted at `/gcp` in main.
///
/// GCP uses the `:action` suffix convention (e.g. `/topics/my-topic:publish`).
/// Axum cannot express two named parameters in a single path segment
/// (`:topic:publish` is invalid syntax), but path parameter extraction captures
/// the full segment including colons, so we register a single POST handler per
/// resource type and dispatch by suffix inside the handler.
pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        // Topics
        .route(
            "/v1/:project/topics/:topic",
            put(create_topic).get(get_topic).delete(delete_topic).post(topic_action),
        )
        .route("/v1/:project/topics", get(list_topics))
        // Subscriptions
        .route(
            "/v1/:project/subscriptions/:sub",
            put(create_subscription)
                .get(get_subscription)
                .delete(delete_subscription)
                .post(subscription_action),
        )
        .route("/v1/:project/subscriptions", get(list_subscriptions))
}

// ── Action dispatchers ────────────────────────────────────────────────────────

/// Dispatch `POST /v1/:project/topics/:topic_action` based on the `:action`
/// suffix of the last path segment (e.g. `my-topic:publish`).
async fn topic_action(
    State(state): State<Arc<AppState>>,
    Path((project, topic_action)): Path<(String, String)>,
    body: axum::body::Bytes,
) -> Response {
    if let Some(topic) = topic_action.strip_suffix(":publish") {
        let json_body: serde_json::Value =
            serde_json::from_slice(&body).unwrap_or(serde_json::json!({}));
        return publish_inner(&state, &project, topic, json_body).await;
    }
    (
        StatusCode::BAD_REQUEST,
        axum::Json(serde_json::json!({ "error": format!("unknown topic action in: {topic_action}") })),
    )
        .into_response()
}

/// Dispatch `POST /v1/:project/subscriptions/:sub_action` based on the `:action`
/// suffix (`:pull` or `:acknowledge`).
async fn subscription_action(
    State(state): State<Arc<AppState>>,
    Path((project, sub_action)): Path<(String, String)>,
    body: axum::body::Bytes,
) -> Response {
    let json_body: serde_json::Value =
        serde_json::from_slice(&body).unwrap_or(serde_json::json!({}));
    if let Some(sub) = sub_action.strip_suffix(":pull") {
        return pull_inner(&state, &project, sub, json_body).await;
    }
    if let Some(sub) = sub_action.strip_suffix(":acknowledge") {
        return acknowledge_inner(&state, &project, sub, json_body).await;
    }
    (
        StatusCode::BAD_REQUEST,
        axum::Json(serde_json::json!({ "error": format!("unknown subscription action in: {sub_action}") })),
    )
        .into_response()
}

// ── Topic handlers ────────────────────────────────────────────────────────────

fn topic_to_json(t: &super::models::PubsubTopic) -> serde_json::Value {
    serde_json::json!({ "name": t.name })
}

fn sub_to_json(s: &super::models::PubsubSubscription) -> serde_json::Value {
    serde_json::json!({ "name": s.name, "topic": s.topic })
}

async fn create_topic(
    State(state): State<Arc<AppState>>,
    Path((_project, topic)): Path<(String, String)>,
) -> Response {
    match svc(&state).create_topic(&topic).await {
        Ok(t) => (StatusCode::OK, Json(topic_to_json(&t))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn get_topic(
    State(state): State<Arc<AppState>>,
    Path((_project, topic)): Path<(String, String)>,
) -> Response {
    match svc(&state).get_topic(&topic).await {
        Ok(Some(t)) => (StatusCode::OK, Json(topic_to_json(&t))).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(serde_json::json!({ "error": "Topic not found" })),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn list_topics(State(state): State<Arc<AppState>>, Path(_project): Path<String>) -> Response {
    match svc(&state).list_topics().await {
        Ok(ts) => {
            let items: Vec<_> = ts.iter().map(topic_to_json).collect();
            (StatusCode::OK, Json(serde_json::json!({ "topics": items }))).into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn delete_topic(
    State(state): State<Arc<AppState>>,
    Path((_project, topic)): Path<(String, String)>,
) -> StatusCode {
    match svc(&state).delete_topic(&topic).await {
        Ok(_) => StatusCode::NO_CONTENT,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

// ── Publish ───────────────────────────────────────────────────────────────────

async fn publish_inner(
    state: &Arc<AppState>,
    _project: &str,
    topic: &str,
    body: serde_json::Value,
) -> Response {
    let msgs: Vec<(String, String)> = body
        .get("messages")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|m| {
                    let data = m
                        .get("data")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    let attrs = m
                        .get("attributes")
                        .map(|v| v.to_string())
                        .unwrap_or_else(|| "{}".to_string());
                    (data, attrs)
                })
                .collect()
        })
        .unwrap_or_default();

    match svc(&state).publish(&topic, msgs).await {
        Ok(ids) => {
            let results: Vec<_> = ids
                .iter()
                .map(|id| serde_json::json!({ "messageId": id }))
                .collect();
            (
                StatusCode::OK,
                Json(serde_json::json!({ "messageIds": results })),
            )
                .into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

// ── Subscription handlers ─────────────────────────────────────────────────────

async fn create_subscription(
    State(state): State<Arc<AppState>>,
    Path((_project, sub)): Path<(String, String)>,
    Json(body): Json<serde_json::Value>,
) -> Response {
    let topic = body.get("topic").and_then(|v| v.as_str()).unwrap_or("");
    match svc(&state).create_subscription(&sub, topic).await {
        Ok(s) => (StatusCode::OK, Json(sub_to_json(&s))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn get_subscription(
    State(state): State<Arc<AppState>>,
    Path((_project, sub)): Path<(String, String)>,
) -> Response {
    match svc(&state).get_subscription(&sub).await {
        Ok(Some(s)) => (StatusCode::OK, Json(sub_to_json(&s))).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(serde_json::json!({ "error": "Subscription not found" })),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn list_subscriptions(
    State(state): State<Arc<AppState>>,
    Path(_project): Path<String>,
) -> Response {
    match svc(&state).list_subscriptions().await {
        Ok(subs) => {
            let items: Vec<_> = subs.iter().map(sub_to_json).collect();
            (
                StatusCode::OK,
                Json(serde_json::json!({ "subscriptions": items })),
            )
                .into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn delete_subscription(
    State(state): State<Arc<AppState>>,
    Path((_project, sub)): Path<(String, String)>,
) -> StatusCode {
    match svc(&state).delete_subscription(&sub).await {
        Ok(_) => StatusCode::NO_CONTENT,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

// ── Pull & Acknowledge ────────────────────────────────────────────────────────

async fn pull_inner(
    state: &Arc<AppState>,
    _project: &str,
    sub: &str,
    body: serde_json::Value,
) -> Response {
    let max = body
        .get("maxMessages")
        .and_then(|v| v.as_i64())
        .unwrap_or(10);
    match svc(&state).pull(&sub, max).await {
        Ok(pairs) => {
            let messages: Vec<_> = pairs
                .iter()
                .map(|(ack, msg)| {
                    serde_json::json!({
                        "ackId": ack.ack_id,
                        "message": {
                            "messageId":  msg.message_id,
                            "data":       msg.data,
                            "attributes": serde_json::from_str::<serde_json::Value>(&msg.attributes)
                                .unwrap_or(serde_json::Value::Object(Default::default())),
                            "publishTime": msg.created_at,
                        }
                    })
                })
                .collect();
            (
                StatusCode::OK,
                Json(serde_json::json!({ "receivedMessages": messages })),
            )
                .into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}

async fn acknowledge_inner(
    state: &Arc<AppState>,
    _project: &str,
    sub: &str,
    body: serde_json::Value,
) -> Response {
    let ack_ids: Vec<String> = body
        .get("ackIds")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default();
    match svc(state).acknowledge(&sub, ack_ids).await {
        Ok(_) => StatusCode::OK.into_response(),
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR.into_response(),
    }
}
